### ISSUE TITLE
how I should receive multipart messages ?

### ISSUE THREAD

**Issue by dainvse04984** [opentelecoms-org/jsmpp#148]
I create smpp server and my client send a long sms with 3 segment. how my smpp server can concat 3 segment to a long message and send to other system ?

**Comment by pmoerenhout** [comment]
Did the client use the optional headers to indicate the segments or did it fill the user data headers ?
You should have an aggregation layer to collect all messages with the same reference. When all parts are received they should be concatenated and forward to the destination if the destination can handle longer messages.

**Comment by dainvse04984** [comment]
jsmpp server support aggregation layer or I have to create it ? I don't no how to create  an aggregation layer to collect all messages.  So you have an idea for this ?

**Comment by pmoerenhout** [comment]
I could image a hashmap for the reference id, and store all message. When all message are collected, a new message can be created. Your server, where does it send the received long message to? Another SMPP link, or an application?

**Comment by dainvse04984** [comment]
> Do you an example of the incoming message(s). I can write an aggregator which will concatenate the messages...

 I have an idea. that i store segment into redis cache. and when get all segment I will concatenate and send to other system.  But I can't appreciate risk if i use it. If you have better idea. I hope you can suggest me. thanks

**Comment by dainvse04984** [comment]
in onAcceptSubmitSm. I got case submitSm.isUdhi(). So how I can get total segment and current segment ?

**Comment by dainvse04984** [comment]
> Do you an example of the incoming message(s). I can write an aggregator which will concatenate the messages...

<img width="1240" alt="Screen Shot 2021-03-18 at 19 39 06" src="https://user-images.githubusercontent.com/27512675/111627550-c11af600-8821-11eb-9189-33de836959d3.png">

**Comment by pmoerenhout** [comment]
OK, hard to parse, but I see no optionalParameters, so the segmentation info is in the user data headers. I have some code to extract it. I will put it together, and come back...

**Comment by pmoerenhout** [comment]
I have made an example in https://github.com/pmoerenhout/pduutils/tree/master/src/main/java/com/github/pmoerenhout/pduutils/jsmpp which parses the user data header of an deliver_sm to a segment. These segments should then be match and the data combined when all segments are collected.

**Comment by dainvse04984** [comment]
> I have made an example in https://github.com/pmoerenhout/pduutils/tree/master/src/main/java/com/github/pmoerenhout/pduutils/jsmpp which parses the user data header of an deliver_sm to a segment. These segments should then be match and the data combined when all segments are collected.

I got your example code and It's work. Thank you very much 💕

**Comment by pmoerenhout** [comment]
Great! I have updated the code to also look at the SMPP OptionalParameters headers to check if SAR information is included.

**Comment by dainvse04984** [comment]
> Great! I have updated the code to also look at the SMPP OptionalParameters headers to check if SAR information is included.

Thank for your update. I'm just a newbie in telecom service and this's my first time I've worked with smpp. So you can give me your contact (Telegram, whatsapp,...) ? I have some question about smpp server-client and I hope you can explain it for me :D
